src/Charcoal/Admin/Widget/QuickFormWidget.php

```diff
@@ -50,6 +85,146 @@ public function submitLabel()
+            if ($locales > 1) {
+                $obj = $this->obj();
+
+                $props = [];
+
+                foreach ($this->groups as $group) {
+                    array_push($props, ...($group->groupProperties()));
+                }
+
+                $props = array_values(array_unique($props));
+
+                foreach ($props as $prop) {
+                    $modelProp = $obj->property($prop);
+
+                    if ($modelProp['l10n']) {
+                        return true;
+                    } elseif ($modelProp instanceof ModelStructureProperty) {
+                        $metadata = $modelProp->getStructureMetadata();
+                        foreach ($metadata->properties() as $prop) {
+                            if (isset($prop['l10n']) && $prop['l10n']) {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        return parent::hasL10nFormProperties();
+    }
+
+    /**
+     * @return boolean
+     */
+    public function showLanguageSwitch()
+    {
+        if ($this->showLanguageSwitch === null) {
+            $this->showLanguageSwitch = $this->hasL10nFormProperties();
+        }
+        return $this->showLanguageSwitch;
+    }
+
+    /**
+     * Determine if content groups are to be displayed as languages tabbable panes.
+     *
+     * @return boolean
+     */
+    public function isDisplayModeLang()
+    {
+        return ($this->groupDisplayMode() === self::DISPLAY_MODE_LANG) ? true : false;
+    }
+
+    /**
+     * Determine if content groups are to be displayed as tabbable panes.
+     *
+     * @return boolean
+     */
+    public function isTabbable()
+    {
+        $tabbableDisplayModes = [
+            self::DISPLAY_MODE_TAB,
+            self::DISPLAY_MODE_LANG
+        ];
+        return in_array($this->groupDisplayMode(), $tabbableDisplayModes) ? true : false;
+    }
+
+    /**
+     * @return array
+     */
+    public function availableLanguages()
+    {
+        $currentLocale = $this->translator()->getLocale();
+        $locales = $this->translator()->locales();
+        $languages = [];
+
+        uasort($locales, [ $this, 'sortLanguagesByPriority' ]);
+
+        foreach ($locales as $locale => $localeStruct) {
+            /**
+             * @see \Charcoal\Admin\Widget\FormGroupWidget::languages()
+             * @see \Charcoal\Property\LangProperty::localeChoices()
+             */
+            if (isset($localeStruct['name'])) {
+                $label = $this->translator()->translation($localeStruct['name']);
+            } else {
+                $trans = 'locale.'.$locale;
+                if ($trans === $this->translator()->translate($trans)) {
+                    $label = strtoupper($locale);
+                } else {
+                    $label = $this->translator()->translation($trans);
+                }
```

Comment:
Charcoal's `Translator#translate()` method converts the input value into a `Translation` object and then returns the localized string. This is only useful if your input value could be a `Translation` object or translation array-shape.

When you know that your input value is a string and all you want is a localized string in return, I recommend using Symfony's `Translator#trans()` method instead:

```suggestion
                $trans = 'locale.'.$locale;
                if ($trans === $this->translator()->trans($trans)) {
                    $label = strtoupper($locale);
                } else {
                    $label = $this->translator()->translation($trans);
                }
```